src/shared_realm.cpp

```diff
@@ -921,6 +941,53 @@ template Object Realm::resolve_thread_safe_reference(ThreadSafeReference<Object>
 template List Realm::resolve_thread_safe_reference(ThreadSafeReference<List> reference);
 template Results Realm::resolve_thread_safe_reference(ThreadSafeReference<Results> reference);
 
+#if REALM_ENABLE_SYNC
+bool Realm::init_permission_cache()
+{
+    if (m_permissions_cache) {
+        if (is_in_transaction())
+            m_permissions_cache->clear();
+        return true;
+    }
+    if (m_config.sync_config && m_config.sync_config->is_partial) {
+        m_permissions_cache = std::make_unique<sync::PermissionsCache>(read_group(), m_config.sync_config->user->identity());
+        return true;
+    }
+    return false;
+}
+void Realm::invalidate_permission_cache()
+{
+    if (m_permissions_cache)
+        m_permissions_cache->clear();
+}
+ComputedPrivileges Realm::get_privileges()
+{
+    if (!init_permission_cache())
+        return ComputedPrivileges::All;
+    return static_cast<ComputedPrivileges>(m_permissions_cache->get_realm_privileges());
+}
+ComputedPrivileges Realm::get_privileges(StringData object_type)
+{
+    if (!init_permission_cache())
+        return ComputedPrivileges::All;
+    return static_cast<ComputedPrivileges>(m_permissions_cache->get_realm_privileges() & m_permissions_cache->get_class_privileges(object_type));
+}
+ComputedPrivileges Realm::get_privileges(RowExpr row)
+{
+    if (!init_permission_cache())
+        return ComputedPrivileges::All;
+    auto& table = *row.get_table();
+    auto object_type = ObjectStore::object_type_for_table_name(table.get_name());
+    sync::GlobalID global_id{object_type, sync::object_id_for_row(read_group(), table, row.get_index())};
+    auto privileges = m_permissions_cache->get_realm_privileges()
+                    & m_permissions_cache->get_class_privileges(object_type)
+                    & m_permissions_cache->get_object_privileges(global_id);
```

Comment:
This looks like a misunderstanding of the underlying semantics. (Though understandable, because this is how it worked in the very initial implementation a few weeks ago.)

Privileges from different levels can't be and'ed together like this, because they don't mean the same things. `Delete` at the object level only requires `Update` at the class and Realm levels, for example.

`Delete` at the class level means the permission to delete the class itself, not the permission to delete objects of that class.

The method `PermissionsCache::can()` was added in order to encapsulate the multilevel logic — however, it only returns a boolean, not the entire bitfield. I'm confused, though, because we went away from using bitfields in Sync due to the inconvenience, but this is turned booleans back into a bitfield?

Reply:
It was, and still is, inconvenient if a bitfield is used in the `__Permission` objects because it makes that object un-queryable, but for the purpose of transporting the calculated permissions to the bindings for exposure in the public API, a bitfield is just as easy to use as exposing specific boolean getters.

Reply:
Bit fields are only problematic as properties of realm objects, as neither core nor the bindings have proper support for them there.